File: warehouse_management/public/js/warehouse.js
frappe.ui.form.on("Warehouse", {
    setup(frm) {
        frm.set_query("main_warehouse", () => {
            return {
                filters: {
                    is_sub_warehouse: 0,
                    disabled: 0,
                },
            };
        });
    },
    warehouse_type(frm) {
		frm.set_value(
			"include_in_transaction",
			["Bay", "Storage"].includes(frm.doc.warehouse_type) ? 1 : 0
		);
	},

	refresh(frm) {
		if (frm.is_new() && frm.doc.warehouse_type) {
			frm.set_value(
				"include_in_transaction",
				["Bay", "Storage"].includes(frm.doc.warehouse_type) ? 1 : 0
			);
		}
	},
});

File: warehouse_management/events/warehouse.py
import frappe
from frappe import _


def validate_warehouse_type(doc, method=None):
    if doc.is_sub_warehouse and doc.warehouse_type == "Storage":
        frappe.throw(
            _("Sub warehouses cannot be marked as Storage")
        )

    if not doc.is_sub_warehouse and doc.warehouse_type in ["Bay", "Hold"]:
        frappe.throw(
            _("Main warehouses cannot be marked as Bay or Hold")
        )

    if doc.warehouse_type in ["Bay", "Storage"]:
        doc.include_in_transaction = 1

    # Only one Outward warehouse is allowed under a main warehouse
    if doc.is_sub_warehouse and doc.warehouse_type == "Outward":
        existing_outward = frappe.db.exists(
            "Warehouse",
            {
                "parent_warehouse": doc.parent_warehouse,
                "warehouse_type": "Outward",
                "is_sub_warehouse": 1,
                "name": ["!=", doc.name],
            },
        )

        if existing_outward:
            frappe.throw(
                _("Only one Outward warehouse is allowed per main warehouse.")
                )